File: Algorithm/go/blind75/242-ValidAnagram.go
package main

func isAnagram(s string, t string) bool {
	if len(s) != len(t) {
		return false
	}

	cache := make(map[rune]int)
	for _, r := range s {
		cache[r]++
	}

	for _, r := range t {
		if v, ok := cache[r]; ok {
			if v-1 < 0 {
				return false
			}
			cache[r] = v - 1
		} else {
			return false
		}
	}

	for _, v := range cache {
		if v != 0 {
			return false
		}
	}

	return true
}

File: Algorithm/python/Blind75/242-ValidAnagram.py
from collections import defaultdict


class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        cache = defaultdict(int)
        for c in s:
            cache[c] += 1

        for c in t:
            if cache[c] - 1 < 0:
                return False
            cache[c] -= 1

        return all(v == 0 for v in cache.values())
